docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,200 @@
+
+The `kubelet` has some support for influencing system behavior in response to a system OOM by
+having the system OOM killer see higher OOM score adjust scores for containers that have consumed
+the largest amount of memory relative to their request.  System OOM events are very compute
+intensive, and can starve the node until the OOM killing process has completed.  In addition,
+the system is prone to return to an unstable state since the containers that are killed due to OOM
+are just restarted, and the node may find itself soon starved of memory again. 
+
+Instead, we would prefer a system where the `kubelet` can pro-actively monitor for
+and prevent against total starvation of a compute resource, and in cases of where it
+could appear to occur, pro-actively fail a pod and terminate its containers so its workload
+can get moved and scheduled elsewhere when/if its backing controller creates a new pod.
+
+## Scope of proposal
+
+This proposal defines a pod eviction policy for reclaiming compute resources.
+
+In the first iteration, it focuses on memory; later iterations are expected to cover
+other resources like disk.  The proposal focuses on a simple default eviction strategy
+intended to cover the broadest class of user workloads.
+
+## Eviction Thresholds
+
+The `kubelet` will support the ability to specify eviction thresholds.  Each threshold 
+defines when a signal has met an administrator defined value that would impact node stability.
+If the threshold criteria has been met, the `kubelet` would take pro-active action to
+attempt to reclaim the starved compute resource associated with the signal.
+
+The `kubelet` will support two types of eviction thresholds.  A soft eviction threshold
+has an associated grace period, such that no action to reclaim the associated resource will
+be taken by the `kubelet` unless the grace period has been exceeded.  A hard eviction
+threshold has no grace period, and if observed, the `kubelet` will take immediate action
+to reclaim the associated starved resource.
+
+Initial support for the feature will focus on available memory, but other potential signals
+that may be added in the future include cpu load or available disk.
+
+If no eviction thresholds are specified, the `kubelet` will not do any eviction.
+
+To configure hard eviction thresholds, the following flag will be supported:
+
+```
+--eviction-hard="": Comma-delimited list of eviction expressions.  For example, 'memory.available<300Mi'
+```
+
+To configure soft eviction thresholds, the following flags will be supported:
+
+```
+--eviction-soft="": Comma-delimted list of eviction expressions.  For example, 'memory.available<500Mi'
+--eviction-soft-period="": Comma-delimited list of eviction periods for each soft eviction signal.  For example, 'memory.available=30s'
+```
+
+If no eviction period is configured for a corresponding soft eviction threshold,
+it will default to 30s for each soft eviction signal.
```

Comment:
@vishh - we could just require the user to provide a value explicitly.

So the following would error:

```
--eviction-soft=memory.available<500Mi
```

because no corresponding `--eviction-soft-period` was provided.

The following would be valid:

```
--eviction-soft=memory.available<500Mi
--eviction-soft-period=memory.available=30s
```

If we configure disk:


```
--eviction-soft=memory.available<500Mi,fs.available<200Mi
--eviction-soft-period=memory.available=30s,fs.available=20s
```

etc.